src/main.cpp

```diff
@@ -1006,39 +981,18 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         }
 
         CTxMemPoolEntry entry(tx, nFees, GetTime(), chainActive.Height(), pool.HasNoInputsOf(tx), fSpendsCoinbase, lp, nSigOps);
-        unsigned int nSize = entry.GetTxSize();
 
         // Don't accept it if it can't get into a block
-        CAmount txMinFee = GetMinRelayFee(tx, nSize, true);
-        if (fLimitFree && nFees < txMinFee)
-            return state.DoS(0, error("AcceptToMemoryPool: not enough fees %s, %d < %d",
-                                      hash.ToString(), nFees, txMinFee),
-                             REJECT_INSUFFICIENTFEE, "insufficient fee");
-
-        // Require that free transactions have sufficient priority to be mined in the next block.
-        if (GetBoolArg("-relaypriority", true) && nFees < ::minRelayTxFee.GetFee(nSize) && !AllowFree(GetPriority(view, tx, chainActive.Height() + 1))) {
-            return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "insufficient priority");
-        }
-
-        // Continuously rate-limit free (really, very-low-fee) transactions
-        // This mitigates 'penny-flooding' -- sending thousands of free transactions just to
-        // be annoying or make others' transactions take longer to confirm.
-        if (fLimitFree && nFees < ::minRelayTxFee.GetFee(nSize))
-        {
-            static double dFreeCount;
-            static int64_t nLastFreeTime;
-            static int64_t nFreeLimit = GetArg("-limitfreerelay", 15);
-            static CCriticalSection csLimiter;
-            LOCK(csLimiter);
-
-            // At default rate it would take over a month to fill 1GB
-            if (RateLimitExceeded(dFreeCount, nLastFreeTime, nFreeLimit, nSize))
-                return state.DoS(0, error("AcceptToMemoryPool : free transaction rejected by rate limiter"),
-                                 REJECT_INSUFFICIENTFEE, "rate limited free transaction");
-
-            LogPrint(Log::MEMPOOL, "Rate limit dFreeCount: %g => %g\n", dFreeCount, dFreeCount+nSize);
+        if (!fLimitFree) {
+            FeeEvaluator feeEval(Opt().AllowFreeTx(), mempool.GetFeeModifier(), ::minRelayTxFee);
+            auto res = feeEval.HasSufficientFee(view, entry, chainActive.Height());
+            if (res != FeeEvaluator::FEE_OK) {
+                std::string err = FeeEvaluator::ToString(res);
+                return state.DoS(0, false, REJECT_INSUFFICIENTFEE, err);
+            }
         }
 
+        unsigned int nSize = entry.GetTxSize();
         if (fRejectAbsurdFee && nFees > ::minRelayTxFee.GetFee(nSize) * 10000)
```

Comment:
Should the absurd-fee logic also be part of FeeEvaluator?

Reply:
Done in a new commit